File: S3/CSL 203 OBJECT ORIENTED PROGRAMMING LAB (IN JAVA)/Model Questions/QuickSort.java
class QuickSort
{
    public static void sort(int arr[])
    {
        quickSort(arr,0, arr.length-1);
    }

    public static void quickSort(int arr[], int l, int h)
    {
        if(l<h)
        {
            int pivot = partition(arr,l,h);
            quickSort(arr, l, pivot-1);
            quickSort(arr, pivot+1, h);
        }
    }

    public static int partition(int arr[], int l, int h)
    {
        int pivot = arr[h];
        int i=l-1;
        for(int j=l; j<h; j++)
        {
            if(arr[j]<pivot)
            {
                i++;
                swap(arr,i,j);
            }
        }
        swap(arr,i+1,h);
        return i+1;
    }

    private static void swap(int arr[], int i, int j)
    {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    public static void display(int arr[])
    {
        for(int i=0; i<arr.length; i++)
            System.out.print("  "+arr[i]);
        System.out.println();
    }

    public static void main(String[] args) {
        int arr[] = {1,0,-2,3,4,-5};
        display(arr);
        sort(arr);
        display(arr);
    }
}

File: S3/CSL 201 DATA STRUCTURES LAB/Model Questions/Sorting/quicksort.c
#include <stdio.h>

int partition(int arr[], int l, int h)
{
    int pivot = arr[h];
    int pIndex = l;

    for(int i=l; i<h; i++)
    {
        if(arr[i]<=pivot)
        {
            int temp = arr[i];
            arr[i] = arr[pIndex];
            arr[pIndex] = temp;

            pIndex++;
        }
    }

    int temp = arr[h];
    arr[h] = arr[pIndex];
    arr[pIndex] = temp;

    return pIndex;
}

void quicksort(int arr[], int l, int h)
{
    if(l<h)
    {
        int j=partition(arr, l, h);
        quicksort(arr, l, j-1);
        quicksort(arr, j+1, h);
    }
}

void main()
{
    int arr[] = {0,2,3,1,4};
    printf("Array: ");
    for(int i=0; i<5; i++)
        printf("%d ",arr[i]);
    quicksort(arr, 0, 5);
    printf("\nSorted array: ");
    for(int i=0; i<5; i++)
        printf("%d ",arr[i]);
}